test/Fencer/Rules/Test.hs

```diff
@@ -71,6 +81,78 @@ test_loadRulesRecursively =
         (sortOn domainDefinitionId [domain1, domain2])
         (sortOn domainDefinitionId definitions)
 
+-- | Test that a rule limit unit change adds a new counter and leaves
+-- the old one intact.
+test_rulesLimitUnitChange :: TestTree
+test_rulesLimitUnitChange =
+  withResource createServerAppState destroyServerAppState $ \ioLogIdState ->
+    testCase "A rule limit unit change on rule reloading" $ do
+      Temp.withSystemTempDirectory "fencer-config-unit" $ \tempDir -> do
+        createDirectoryIfMissing True (tempDir </> dir)
+
+        definitions1 <- writeLoad tempDir merchantLimitsText1
+        (_, _, state) <- ioLogIdState
+
+        atomically $ setRules state (mapRuleDefs definitions1)
+
+        ruleTree :: RuleTree <- atomically $
+          fromMaybe' <$> StmMap.lookup domainId (appStateRules state)
+        let ruleBranch = fromMaybe' $ HM.lookup (ruleKey, Just ruleValue) ruleTree
+        let rateLimit =  fromMaybe' $ ruleBranchRateLimit ruleBranch
+
+        -- Record a hit
+        void $ atomically $ recordHits state (#hits 1) (#limit rateLimit) counterKey1
+
+        mV1 <- atomically $ StmMap.lookup counterKey1 $ appStateCounters state
+
+        -- Change rules in the configuration
+        definitions2 <- writeLoad tempDir merchantLimitsText2
+
+        -- Set the new rules and the rules reloaded flag
+        atomically $ setRules state (mapRuleDefs definitions2)
+
+        mV1' <- atomically $ StmMap.lookup counterKey1 $ appStateCounters state
+        mV2  <- atomically $ StmMap.lookup counterKey2 $ appStateCounters state
+
+        assertBool
+          "The original counter was not updated after recording a hit!"
+          ((counterHits <$> mV1) == Just 1)
+        assertBool
+          "The original counter was mistakenly updated in the meantime!"
+          (mV1 == mV1')
+        assertBool "The secondary counter was set!" (mV2 == Nothing)
+ where
+  mapRuleDefs :: [DomainDefinition] -> [(DomainId, RuleTree)]
+  mapRuleDefs defs =
+    [ ( domainDefinitionId rule
+      , definitionsToRuleTree (NE.toList . domainDefinitionDescriptors $ rule))
+    | rule <- defs
+    ]
+
+  dir     = "d11-ratelimits"
+  cfgFile = "d11-ratelimits1.yaml"
+
+  writeLoad :: FilePath -> Text -> IO [DomainDefinition]
```

Comment:
Since we are already testing `loadRulesFromDirectory` independently, I think that in this test we can assume it is trustworthy, and encode rulesets as `DomainDefinition`s instead of going via YAML.